docs/hacking.rst

```diff
@@ -2,273 +2,256 @@
 
-A single localization unit in L20n vocabulary is called an entity.
-An entity is an ``Object`` with three fields:
+A single localization unit in ``MessageContext`` is called an entity.  An entity 
+has an identifier, a value and traits.
 
-:id:
-    An identifier
+:id
+    The unique identifier of the entity.
 
-:value:
-    A value that may be a simple or more complex computable string.
+:value
+    The entity's value. It may be null.
 
-:traits:
+:traits
     A list of key-value pairs that can be providing additional data for the entity like
-    attributes or variants.
+    attributes or variants. It may be empty.
+
+Entities are encoded in a source format named FTL. l20n.js has two FTL parsers.
+The full AST parser is available for tools and external consumers who need
+access to the full AST.  For runtime purposes, l20n.js uses a smaller and 
+faster Runtime parser, which has many optimizations that target performance and 
+smaller memory footprint.
 
-Entities are encoded in a source format named FTL. L20n provides two parsers for the
-format. The full AST parser is available for tools and external users who need
-access to the full AST of the file.
-For runtime purposes, L20n uses a smaller and faster Entries parser, which
-has many optimizations that target performance and smaller memory footprint.
 
 MessageContext
 --------------
 
-``MessageContext`` is a synchronous single-locale Map and doesn't provide any fallback.
-
-It parses source string into a list of messages, and allows the user to retrieve
-``Entity`` objects formatted using environmental variables, user provided arguments,
-other messages in the context and any syntax that the format allows for.
-
-The entities are parsed into a lazy intermittent form and then formatted
-and cached only on retrieval.
-
+``MessageContext`` is a synchronous single-locale map of translations.  It can 
+format translations to strings.
 
-``MessageContext`` has two methods:
+The translations can be added with the ``addMessages`` method.  They are 
+internally parsed into a lazy intermittent form and then formatted only on 
+retrieval.  The ``format`` method allows the user to format them using 
+environmental variables, user-provided arguments, other messages in the current 
+context and any other syntax that the format allows.
 
-:constructor(lang, options):
-    lang
-        a language tag string
-    options 
-        functions
-            an object with environmental function that are going to be available
-            from within the entities.
-    
-:addMessages(source):
-    takes a localization file source (at the moment FTL file format is supported),
-    parses the source into localization messages and stores them.
+The translations are considered untrusted. ``MessageContext`` has been 
+designed to be resilient to formatting errors.  If a translation contains 
+errors, ``MessageContext`` will try to salvage as much of the translation as 
+possible and will still return a string.
```

Comment:
This reads a bit weird. The `The translations are considered untrusted` being the first sentence of the paragraph feels odd.
Maybe put it further down in the paragraph or explain a bit: `L20n takes an approach of assuming that localization resources do not go through a security review process and may be coming from a third-party, thus the translation have to be considered untrusted`.